buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java

```diff
@@ -61,6 +61,10 @@
     private final File artifactsExtractDir;
     private final LinkedHashMap<String, Predicate<TestClusterConfiguration>> waitConditions = new LinkedHashMap<>();
     private final GradleServicesAdapter services;
+    // We have a rate limiting mechanism to make sure we don't have too many nodes running at the same time in the build
+    // We use this flag to signal when the cluster is actually started and we should start the counter and wait for it
+    // to come online. We just block if sufficient resources are not available.
+    private final AtomicBoolean shouldWait = new AtomicBoolean(false);
```

Comment:
The restart operation on the cluster is implemented as a restart on all nodes. It doesn't call `stop` and `start` on the cluster itself. That's why we also  need the boolean flag. Once the cluster is started it will stay `true`, we never flip it back, so waiting for the cluster will proceed to checking ports file and cluster health without blocking on the lock.